File: leet_code/linked_lists/largest_substring.rb
def largest_substring(string)
  return nil if !string
  return 1 if string.length == 1

  longest, left, right = 0, 0, 0
  count = Hash.new

  while left < string.length
    if right >= string.length || count[string[right]]
      longest = (right - left) if (right - left) > longest
      left += 1
      right = left
      count = Hash.new
    else
      count[string[right]] = 1
      right += 1
    end
  end
  longest
end

File: leet_code/linked_lists/largest_substring.js
const largest_substring = string => {
  if (!string) return null;
  if (string.length === 1) return 1;

  let longest = 0,
      left = 0,
      right = 0;

  let count = {};

  while (left < string.length) {
    if (right >= string.length || count[string[right]]) {
      if (right - left > longest) longest = right - left;
      left++;
      right = left;
      count = {};
    } else {
      count[string[right]] = 1;
      right++;
    }
  }
  return longest;
};
